lib/backend/pgbk/pgbk.go

```diff
@@ -451,61 +480,77 @@ func (b *Backend) Get(ctx context.Context, key []byte) (*backend.Item, error) {
-		// TODO(espadolini): figure out if we want transaction_deferred enabled
-		// for GetRange
-
-		var items []backend.Item
-		batch.Queue(
-			"SELECT kv.key, kv.value, kv.expires, kv.revision FROM kv"+
-				" WHERE kv.key BETWEEN $1 AND $2 AND (kv.expires IS NULL OR kv.expires > now())"+
-				" ORDER BY kv.key LIMIT $3",
-			nonNil(startKey), nonNil(endKey), limit,
-		).Query(func(rows pgx.Rows) error {
-			var err error
-			items, err = pgx.CollectRows(rows, func(row pgx.CollectableRow) (backend.Item, error) {
-				var key, value []byte
-				var expires time.Time
-				var revision revision
-				if err := row.Scan(&key, &value, (*zeronull.Timestamptz)(&expires), &revision); err != nil {
-					return backend.Item{}, err
-				}
-				return backend.Item{
-					Key:      key,
-					Value:    value,
-					Expires:  expires.UTC(),
-					ID:       idFromRevision(revision),
-					Revision: revisionToString(revision),
-				}, nil
+	lastSeenKey := startKey
+	results := &backend.GetResult{}
+	pageSize := defaultPageSize
+
+	if limit > 0 && pageSize > limit {
+		pageSize = limit
+	}
+
+	for {
+		items, err := pgcommon.RetryIdempotent(ctx, b.log, func() ([]backend.Item, error) {
+			batch := new(pgx.Batch)
+			// batches run in an implicit transaction
+			batch.Queue("SET transaction_read_only TO on")
+			// TODO(espadolini): figure out if we want transaction_deferred enabled
+			// for GetRange
+
+			var items []backend.Item
+			batch.Queue(
+				"SELECT kv.key, kv.value, kv.expires, kv.revision FROM kv"+
+					" WHERE kv.key BETWEEN $1 AND $2 AND (kv.expires IS NULL OR kv.expires > now())"+
+					" ORDER BY kv.key LIMIT $3",
+				NonNil(lastSeenKey), NonNil(endKey), pageSize,
+			).Query(func(rows pgx.Rows) error {
+				var err error
+				items, err = pgx.CollectRows(rows, func(row pgx.CollectableRow) (backend.Item, error) {
+					var key, value []byte
+					var expires time.Time
+					var revision revision
+					if err := row.Scan(&key, &value, (*zeronull.Timestamptz)(&expires), &revision); err != nil {
+						return backend.Item{}, err
+					}
+					return backend.Item{
+						Key:      key,
+						Value:    value,
+						Expires:  expires.UTC(),
+						ID:       IdFromRevision(revision),
+						Revision: RevisionToString(revision),
+					}, nil
+				})
+				return trace.Wrap(err)
 			})
-			return trace.Wrap(err)
-		})
 
-		if err := b.pool.SendBatch(ctx, batch).Close(); err != nil {
+			if err := b.pool.SendBatch(ctx, batch).Close(); err != nil {
+				return nil, trace.Wrap(err)
+			}
+			return items, nil
+		})
+		if err != nil {
 			return nil, trace.Wrap(err)
 		}
-
-		return items, nil
-	})
-	if err != nil {
-		return nil, trace.Wrap(err)
+		results.Items = append(results.Items, items...)
+		if len(items) < pageSize {
+			break
+		}
+		if limit > 0 && len(results.Items) >= limit {
+			break
+		}
+		if limit > 0 && len(results.Items)+pageSize > limit {
+			pageSize = limit - len(results.Items)
+		}
+		lastSeenKey = items[len(items)-1].Key
```

Comment:
`BETWEEN` is inclusive, so this will result in a duplicated item at the end of each page boundary. The correct way to fix this would be to tweak the query for the fetches after the first to be `$1 < key AND key <= $2` instead of `key BETWEEN $1 AND $2`; we could also use `backend.NextKey` here tho.